### PR TITLE
Use HashSet to implement dedupArrayList to improve time complexity

### PR DESCRIPTION
**Pull request by zhiming-qiu**
Use a `HashSet` to detect already appeared items and hence avoid using two loops (O(n^2)).

### CODE REVIEW

**Comment by zhiming-qiu** [comment]
@EachOneChew Pls take a look. Note I choose `HashSet` since it is O(1) to check if it has an element. If using other types such as `TreeSet`, the complexity drops to O(log(n)).

**Comment by zhiming-qiu** [comment]
@EachOneChew Actually it is better to perform the check when we add stuff into `toVisit` instead of adding duplicates and then performing dedup later on. This is a future issue though.

File: src/DistanceCalculator.java [hunk 1]
```diff
@@ -35,6 +36,18 @@
         {0,  0,  0,  0, -1,  0} 
     };
 
+    private int getMoveRangeFromMoveType(char moveType) {
```
Comment (zhiming-qiu) [inline src/DistanceCalculator.java:39]: By making this helper method, you can save a few lines of typing. :-)

**Comment by zhiming-qiu** [comment]
I noticed an issue. Here below is the output of your main program:

``
2          1          2          3          4          5

1          0          1          4          5          6

2147483647 2147483647 2147483647 5          2147483647 7

9          8          7          6          7          8

10         9          8          7          8          9

11         10         9          8          9          10

12         11         10         9          10         11

13         12         11         10         11         12
``

In the second row, however, the fourth cell should have value 2 instead of 4, shouldn't it?

**Review by EachOneChew — APPROVED** [review]
Makes sense.

The distance at (3, 1) is intended: there is an enemy - target4 at (2, 1) - between the origin point and (3, 1), so you actually need to loop above in a path since you can't walk through enemies.

See DistanceCalculator where I set delays to -2 and how -2 delays are handled.

**Pull request merged**